doc/manual/performance-tips.rst

```diff
@@ -107,64 +69,167 @@ version of your function that separates allocation from its
+For example, the following contrived function returns an array of a
+randomly-chosen type::
+
+    function strange_twos(n)
+        a = Array(rand(Bool) ? Int64 : Float64, n)
+        for i = 1:n
+            a[i] = 2
+        end
+        return a
     end
 
-Because ``a`` is a an array of abstract type :class:`Real`, it must be able
-to hold any Real value.  Since :class:`Real` objects can be of arbitrary
-size and structure, ``a`` must be represented as an array of pointers to
-individually allocated :class:`Real` objects.  Because ``f`` will always be
-a :class:`Float64`, we should instead, use::
+This should be written as::
 
-    a = Float64[] # typeof(a) = Array{Float64,1}
+    function fill_twos!(a)
+        for i=1:length(a)
+            a[i] = 2
+        end
+    end
+
+    function strange_twos(n)
+        a = Array(rand(Bool) ? Int64 : Float64, n)
+        fill_twos!(a)
+        return a
+    end
+
+Julia's compiler specializes code for argument types at function
+boundaries, so in the original implementation it does not know the type
+of ``a`` during the loop (since it is chosen randomly). Therefore the
+second version is generally faster since the inner loop can be
+recompiled as part of ``fill_twos!`` for different types of ``a``.
+
+The second form is also often better style and can lead to more code
+reuse.
+
+This pattern is used in several places in the standard library. For
+example, see ``hvcat_fill`` in
+`abstractarray.jl <https://github.com/JuliaLang/julia/blob/master/base/abstractarray.jl>`_,
+or the ``fill!`` function, which we could have used instead of writing
+our own ``fill_twos!``.
+
+Functions like ``strange_twos`` occur when dealing with data of
+uncertain type, for example data loaded from an input file that might
+contain either integers, floats, strings, or something else.
+
+Break functions into multiple definitions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Writing a function as many small definitions allows the compiler to
+directly call the most applicable code, or even inline it.
+
+Here is an example of a "compound function" that should really be
+written as multiple definitions::
+
+    function norm(A)
+        if isa(A, Vector)
+            return sqrt(real(dot(A,A)))
+        elseif isa(A, Matrix)
+            return max(svd(A)[2])
+        else
+            error("norm: invalid argument")
+        end
+    end
+
+This can be written more concisely and efficiently as::
+
+    norm(x::Vector) = sqrt(real(dot(x,x)))
+    norm(A::Matrix) = max(svd(A)[2])
 
-which will create a contiguous block of 64-bit floating-point values
-that can be manipulated efficiently.
 
-See also the discussion under :ref:`man-parametric-types`.
+
+Inline functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+While it's always a good idea to define multiple functions depending on a type,
+be careful when calling a function within a loop: functions are not always inlined.
+
```

Comment:
This sounds like a dire problem, but you don't want to encourage rampant `@inline`. Indeed, the tricks in the "Separate kernel functions" section would fail were you to put `@inline` in front of them: inlining destroys the "function barrier" that allows julia to make a type-stable realization of the loop.

See also #12219.